flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/ArrayExceptFunction.java

```diff
@@ -0,0 +1,102 @@
+import org.apache.flink.table.functions.BuiltInFunctionDefinitions;
+import org.apache.flink.table.functions.FunctionContext;
+import org.apache.flink.table.functions.SpecializedFunction;
+import org.apache.flink.table.runtime.util.EqualityAndHashcodeProvider;
+import org.apache.flink.table.runtime.util.ObjectContainer;
+import org.apache.flink.table.types.CollectionDataType;
+import org.apache.flink.table.types.DataType;
+import org.apache.flink.util.FlinkRuntimeException;
+
+import javax.annotation.Nullable;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/** Implementation of {@link BuiltInFunctionDefinitions#ARRAY_EXCEPT}. */
+@Internal
+public class ArrayExceptFunction extends BuiltInScalarFunction {
+    private final ArrayData.ElementGetter elementGetter;
+    private final EqualityAndHashcodeProvider equalityAndHashcodeProvider;
+
+    public ArrayExceptFunction(SpecializedFunction.SpecializedContext context) {
+        super(BuiltInFunctionDefinitions.ARRAY_EXCEPT, context);
+        final DataType dataType =
+                ((CollectionDataType) context.getCallContext().getArgumentDataTypes().get(0))
+                        .getElementDataType()
+                        .toInternal();
+        elementGetter = ArrayData.createElementGetter(dataType.toInternal().getLogicalType());
+        this.equalityAndHashcodeProvider = new EqualityAndHashcodeProvider(context, dataType);
+    }
+
+    @Override
+    public void open(FunctionContext context) throws Exception {
+        equalityAndHashcodeProvider.open(context);
+    }
+
+    public @Nullable ArrayData eval(ArrayData arrayOne, ArrayData arrayTwo) {
+        try {
+            if (arrayOne == null || arrayTwo == null) {
+                return null;
+            }
+
+            List<Object> list = new ArrayList<>();
+            Map<ObjectContainer, Integer> map = new HashMap<>();
+            for (int pos = 0; pos < arrayTwo.size(); pos++) {
+                final Object element = elementGetter.getElementOrNull(arrayTwo, pos);
+                if (element == null) {
+                    map.merge(null, 1, (k, v) -> v + 1);
+                } else {
+                    ObjectContainer objectContainer = new ObjectContainer(element);
+                    map.merge(objectContainer, 1, (k, v) -> v + 1);
+                }
+            }
+            for (int pos = 0; pos < arrayOne.size(); pos++) {
+                final Object element = elementGetter.getElementOrNull(arrayOne, pos);
+                final ObjectContainer objectContainer =
+                        element == null ? null : new ObjectContainer(element);
+                if (map.containsKey(objectContainer)) {
+                    if (map.get(objectContainer) > 1) {
+                        map.put(objectContainer, map.get(objectContainer) - 1);
+                    } else {
+                        map.remove(objectContainer);
+                    }
```

Comment:
Let's be consistent. If you do here:
```
final ObjectContainer objectContainer = element == null ? null : new ObjectContainer(element);
```
do it also few lines above:
```
                if (element == null) {
                    map.merge(null, 1, (k, v) -> v + 1);
                } else {
                    ObjectContainer objectContainer = new ObjectContainer(element);
                    map.merge(objectContainer, 1, (k, v) -> v + 1);
                }
```

Don't mix and match different approaches, because it only confuses the reader. Code is read more often than written and we should optimise for reading. Doing the same thing in two different ways forces the reader to analyze the code twice spending twice as much time reviewing it.

Reply:
ok, I will use few lines.